vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -364,83 +381,115 @@ def register_kv_caches(self, kv_caches: dict[str, torch.Tensor]):
-            for block_id in range(num_blocks):
-                block_offset = block_id * dst_block_len
-                blocks_data.append((base_addr + block_offset, dst_block_len,
-                                    self.rank * tp_multiplier))
+            for block_id in range(nixl_agent_meta.num_blocks):
+                block_offset = block_id * self.block_len
+                # (addr, len, device id)
+                blocks_data.append(
+                    (base_addr + block_offset, self.block_len, self.rank))
         logger.debug("Created %s blocks for dst engine %s and rank %s",
                      len(blocks_data), engine_id, self.rank)
+
         # Register with NIXL.
         descs = self.nixl_wrapper.get_xfer_descs(blocks_data, "VRAM")
-        self.dst_xfer_side_handles[engine_id][tp_idx] = (
-            self.nixl_wrapper.prep_xfer_dlist(
-                self._remote_agents[engine_id][self.rank * tp_multiplier +
-                                               tp_idx], descs))
+        self.dst_xfer_side_handles[
+            engine_id] = self.nixl_wrapper.prep_xfer_dlist(
+                self._remote_agents[engine_id], descs)
 
     def get_finished(self) -> tuple[set[str], set[str]]:
         """Get requests that are done sending or recving."""
         done_sending = self._get_new_notifs()
         done_recving = self._pop_done_transfers(self._recving_transfers)
         if len(done_sending) > 0 or len(done_recving) > 0:
             logger.debug(
-                "get_finished: %s requests done sending "
-                "and %s requests done recving", len(done_sending),
+                "Rank %s, get_finished: %s requests done sending "
+                "and %s requests done recving", self.rank, len(done_sending),
                 len(done_recving))
-        return done_sending, done_recving
+
+        if self.world_size == 1:
+            return done_sending, done_recving
+
+        # In TP>1 setup, each rank exchanges KVs with its counterpart
+        # ranks independently. get_finished() runs in a worker creates
+        # the done_sending and done_recving sets that are sent to the
+        # scheduler via ModelRunnerOutput by Rank 0. To avoid race
+        # ensure trnxs are done before adding to finished, Ranks 1 to
+        # N-1 communicate to Rank 0 once their transaction is done.
+        # Rank 0 only returns finished once all ranks are complete.
+        if self.rank == 0:
+            for req_id in done_sending:
+                self._done_sending_count[req_id] += 1
+            for req_id in done_recving:
+                self._done_recving_count[req_id] += 1
+
+            # Update the counts of how many ranks have finished.
+            # Get notifies from other ranks that txns are done.
+            other_ranks_finished_ids: list[str] = []
+            for i in range(1, self.world_size):
+                other_ranks_finished_ids.extend(
+                    self.tp_group.recv_object(src=i))
```

Comment:
this is how Dyanmo does it (with the tp_group)

I wonder if there is a better way cc @njhill

Reply:
@robertgshaw2-redhat here is an alternative to consider https://github.com/robertgshaw2-redhat/vllm/pull/7

Guess this might be preferable latency wise since we don't have additional gather collective, but not sure (since now scheduler needs to receive from all ranks .. though it was doing this anyhow until recently).

Reply:
lets just time things and see which one is faster

Reply:
Looks like for TP=2, the setup I have it taking <1ms, so I think this is good enough for now